### ISSUE TITLE
More puzzles

### ISSUE THREAD

**Issue by Anders-Holst** [SteffenBauer/sgtpuzzles-extended#4]
Hi @SteffenBauer 

This is not really an issue - I am just trying to get in touch with you to ask if you are interested to add a couple of more puzzles.

Some years ago I developed a few puzzles for the sgt-puzzles collection, but for various reasons they didn't make it into the official collection. Which is a pity because I think they are quite good. The most notable one is Kakuro - I don't think any puzzle collection would be complete without it. Then I stumbled upon this repository by chance. So I though that, instead of everyone making their own repositories with add-ons to sgt-puzzles, there might be a point trying to pool them.

Please let me know if you want to know more. Either here, or you can mail me: anders dot holst at ri dot se.
(By the way, I'm into machine learning too.)

Best Regards
  Anders Holst

**Comment by SteffenBauer** [comment]
Hi Anders,

thanks for the suggestion. I'm definitely interested in your puzzles; also I'm on the lookout for new puzzles to add to my port of the puzzle collection to the PocketBook eReaders.

But as you already have a GitHub account, I'd rather suggest that you put your puzzles into a new repository in your account. First you are the main author, so credit should get to you; furthermore a GitHub account with interesting projects is a very good reference for job hunting.

I would therefore not pool all inofficial puzzles in once location. Simon Tathams repository still is the one official location, but he seems to be reluctant to add puzzles there; it also happened to me that he for whatever reason does not like some of my inofficial extensions. I would rather keep all inofficial extensions connected with each individual author.

**Comment by SteffenBauer** [comment]
On a related note; in case you did not find it yet, there is also another account with a significant amount of puzzles at Lennard Sprongs repository:

https://github.com/x-sheep/puzzles-unreleased

**Comment by Anders-Holst** [comment]
Thanks for your reply. I can see your point. And of course the freedom to decide oneself when something is sufficient quality for release is an advantage of separating the repositories.

On the other hand, from the user perspective it may be confusing with many different repositories to add separately. And add subdirectories "extensions", "unreleased", etc. What can I come up with? "Supplemental"? And what is the difference between these? How can the user tell which are good quality and which are crackpots? (An alternative naming would be "SteffenBauerPuzzles", "LennardSprongPuzzles", "AndersHolstPuzzles", etc, to make more clear that these are individual initiatives.) That's why my first impulse was to see if we can pool efforts, to collect *serious* alternatives to Sgt at one place with one installation instruction, since he, as you say, seems reluctant to add some puzzles. 

(In my case I think it was partly miscommunication. Kakuro is very difficult to generate. At least a couple of years ago (I did not check recently) there were no free, good, generators. It was claimed at one site that the complexity of generating Kakuro is NP-complete in the size. Many other types of puzzles can be generated randomly and then checked if the solution is unique. For an arbitrary randomly generated Kakuro board, the number of solutions are in the orders of 100 millions. No chance in the world to stumble upon one with a unique solution by chance. So I had to rely on my machine learning knowledge, or in more specific: I used an evolutionary process to mutate a Kakuro board to decrease the number of unique solutions until only one remained. In the beginning it took unfeasible long time, over an hour for a small board, and larger boards quite impossible. So I tweaked and optimized, and increased the speed several magnitudes, so that a small board (7x7) was generated in much less than a second, and even a very large board (12x12), previously impossible, was now generated in about 15 seconds.
Then Sgt comments: "No other game has required ML to generate" (between the lines - it is a bad thing and should no be needed. But brute force searching would not be an alternative here.), and "It is much slower to generate than the other games" (the largest suggested size, yes, but we can limit the suggested sizes further. And  it is much more difficult to generate than other games) "So your code is not of enough quality, sorry" (well, I have been teaching programming, I have a long history of contributing to free ware projects, even got code into the Emacs distribution, and care about high quality code, and no-one ever called my code bad-quality before...) So, no sense in pushing further with him.)

Well, what I will initially is to fork your repository (or Leonards - besides you two there seem to be only stray additions, or various ports, so those two are the "serious" attempts of complementing Sgt in my mind), and add my games there, and keep you updated (may take some time). And if you have the time to check it then, it would be nice to know what you think, and if some of it suits your eReader port. (Even criticizing the quality is ok of course - I think I'm rather reasonable, and constructive critique and feedback always makes the code better. And programming as its best is a collective endeavor.) 

 / Anders

**Comment by x-sheep** [comment]
`puzzles-unreleased` is a strange title, because I originally used it to place puzzles that were also turned down by Simon, back when I was very much an amateur at writing C programs. Since then, I've used it mostly to place experimental generators that I believe produce interesting results. The source code contains several rough edges, and a few of the puzzles are blatantly unfinished (and aren't included in the PocketPuzzles port for that reason). I definitely wouldn't call it a serious attempt at making my own collection 😅

I'm interested in seeing your Kakuro implementation, but I agree with Steffen that it's best if you put it in a repository under your own name.

There's actually multiple puzzles in Simon's repository that only partially generate a valid grid, and then regenerates ambiguous areas. Some generators also identify common patterns leading to ambiguous puzzles, then makes a small tweak to resolve the problem.

I'm curious to see how it differs from the Machine Learning mentioned in your post, and how your generator would perform on a device with very limited memory and CPU power like an eReader.

**Comment by Anders-Holst** [comment]
Hi Lennard,
Ok, so I have to admit that I did not have time to try any of those puzzles yet, but I definitely will! However, it appears you got some of them included in the Android port, so at least some of them have passed "peer review" :smiley: 

I'll put them up and let you know. Yes it would be interesting to know how it performs.

I like the principle you both use, of installing sgt-puzzles first and then adding the extras in a subdirectory and compile it all at once. So I'll stick to that too for now. Also I start with a fork, because its easier than starting from scratch...

**Comment by Anders-Holst** [comment]
Hi again, @SteffenBauer and @x-sheep 

(Sorry for reusing this "issue" for posting. I may use my own page in the future...)

I have now cleaned up, completed, documented, and created a git repository for my additional games. Its here: [sgt-puzzles-aho-extensions](https://github.com/Anders-Holst/sgt-puzzles-aho-extensions/) You are welcome to check them out. I reused the same format as your pages use, as you will notice. 

I greatly enjoyed both of yours games! 
Steffen, I especially like creek, kropki, and walls. You noticed I suggested a bugfix for creek.
And Lennard, I like *all* of yours. Even Seismic and Crossing - why did you abandon them? I especially like Boats. You will notice that I have my own variant of Battle ships on my page, called Identifier. The same idea to start from but quite different in its game play.

Best,
  Anders
